### DISCUSSION TITLE
2021.07 - Impractical Information

### DISCUSSION THREAD

**Discussion by hugsy** [hugsy/gef#672] [GEF Releases]
# Changelog: 2021.07 - Impractical Information

## Highlights of `Impractical Information`

 * New unified command argument parser
 * Better aliasing (thanks @hoefler02)
 * More documentation (thanks @plan-do-break-fix )
 * Updated the `pattern` command to strictly follow its usage in `pwntools`
 * Bug fixes in AARCH64 (thanks @borismol )
 * Improved CI & CD (Actions, Pylint, regressions)
 * Bug fix in `heap bin tcache` for multi-threaded execution (thanks @irontigran )
 * Started moving support for IDA/Binary Ninja interface to `gef-extras`, by leveraging `rpyc`

## Contributors

| Name | Number of commits | 
|--|--| 
| Boris Molodenkov | 6 |
| Grazfather | 4 |
| hugsy | 79 |
| Jacob Heybey | 1 |
| Jeffrey Swan | 1 |
| Michael Hoefler | 14 |
| Simon Wörner | 1 |

## Closed Issues

 * 12 issues closed ( [664](https://github.com/hugsy/gef/issues/664) &bull;  [663](https://github.com/hugsy/gef/issues/663) &bull;  [660](https://github.com/hugsy/gef/issues/660) &bull;  [659](https://github.com/hugsy/gef/issues/659) &bull;  [658](https://github.com/hugsy/gef/issues/658) &bull;  [657](https://github.com/hugsy/gef/issues/657) &bull;  [655](https://github.com/hugsy/gef/issues/655) &bull;  [653](https://github.com/hugsy/gef/issues/653) &bull;  [648](https://github.com/hugsy/gef/issues/648) &bull;  [647](https://github.com/hugsy/gef/issues/647) &bull;  [643](https://github.com/hugsy/gef/issues/643) &bull;  [634](https://github.com/hugsy/gef/issues/634) &bull; )

## Closed Pull Requests

 * 18 PRs closed ( [671](https://github.com/hugsy/gef/pull/671) &bull;  [670](https://github.com/hugsy/gef/pull/670) &bull;  [669](https://github.com/hugsy/gef/pull/669) &bull;  [668](https://github.com/hugsy/gef/pull/668) &bull;  [666](https://github.com/hugsy/gef/pull/666) &bull;  [665](https://github.com/hugsy/gef/pull/665) &bull;  [662](https://github.com/hugsy/gef/pull/662) &bull;  [661](https://github.com/hugsy/gef/pull/661) &bull;  [656](https://github.com/hugsy/gef/pull/656) &bull;  [652](https://github.com/hugsy/gef/pull/652) &bull;  [649](https://github.com/hugsy/gef/pull/649) &bull;  [644](https://github.com/hugsy/gef/pull/644) &bull;  [640](https://github.com/hugsy/gef/pull/640) &bull;  [639](https://github.com/hugsy/gef/pull/639) &bull;  [638](https://github.com/hugsy/gef/pull/638) &bull;  [637](https://github.com/hugsy/gef/pull/637) &bull;  [635](https://github.com/hugsy/gef/pull/635) &bull;  [633](https://github.com/hugsy/gef/pull/633) &bull; )

## Commit details

<details><summary>106 commits since <b>2021.04</b></summary>

## Commits

 * 2021-04-07 [013ba5e](http://github.com/hugsy/gef/commit/013ba5e2115e21d3587a0563598b5f63a4ef981d) &bull; *Michael Hoefler* &bull; Added alias options: add, remove, list 
 * 2021-04-07 [3477d39](http://github.com/hugsy/gef/commit/3477d39efda169475eceda736d7249a12e6f5c2b) &bull; *Michael Hoefler* &bull; fixed typo 
 * 2021-04-12 [29368d0](http://github.com/hugsy/gef/commit/29368d04a03d4a297a63dd43294919b1cd62fe68) &bull; *Michael Hoefler* &bull; subcommands / tab completion 
 * 2021-04-12 [46b24aa](http://github.com/hugsy/gef/commit/46b24aab32a051aeedec343a039dd14bf6144af8) &bull; *Michael Hoefler* &bull; clearer code 
 * 2021-04-12 [1cb15f4](http://github.com/hugsy/gef/commit/1cb15f4990df85322bfc321135ddd28d3cc88062) &bull; *Michael Hoefler* &bull; update docs for aliases command 
 * 2021-04-14 [df34985](http://github.com/hugsy/gef/commit/df3498589b1b27790837b15a4fb0a42a7103a968) &bull; *Michael Hoefler* &bull; suggested changes 
 * 2021-04-14 [908db30](http://github.com/hugsy/gef/commit/908db309d7eb347a4dbd08ee5ee26aa68b109410) &bull; *Michael Hoefler* &bull; unit test 
 * 2021-04-16 [0ad5f99](http://github.com/hugsy/gef/commit/0ad5f9976cb09071ff83456e1c285e80f9e02b6a) &bull; *Jacob Heybey* &bull; Upgrade the tcache command to work for all threads 
 * 2021-04-28 [5ff0bdc](http://github.com/hugsy/gef/commit/5ff0bdc27aa9688b4fdd16c424d8fa0ea040274b) &bull; *Jeffrey Swan* &bull; Fix some typos in the docs 
 * 2021-05-18 [2479d6a](http://github.com/hugsy/gef/commit/2479d6a9bd5454b566d71576a42d821482c99455) &bull; *Grazfather* &bull; ida_gef: Add api map for DelBpt 
 * 2021-05-21 [4ae9dbb](http://github.com/hugsy/gef/commit/4ae9dbbbca1623787ab88a378e17bbfdec99727b) &bull; *Simon Wörner* &bull; fix address underflow on disassemble (#652) 
 * 2021-06-05 [c032bf6](http://github.com/hugsy/gef/commit/c032bf6b5f1e4240a0e6fc3a00869223d0797dcd) &bull; *Boris Molodenkov* &bull; Fix aarch64 tbz, tbnz, cbz and cbnz instructions estimation 
 * 2021-06-06 [fac0fb4](http://github.com/hugsy/gef/commit/fac0fb46fa0d162ad53aee78be235c8e064db298) &bull; *Michael Hoefler* &bull; minor changes 
 * 2021-06-07 [03bb804](http://github.com/hugsy/gef/commit/03bb804c4ea5a7ce03351987cf0a78c270141bb4) &bull; *Michael Hoefler* &bull; remove automatic saves 
 * 2021-06-08 [aa38365](http://github.com/hugsy/gef/commit/aa38365b495facb51583015679c9596a26a782ae) &bull; *hugsy* &bull; implemented argparsing using argparse, applied on unicorn-emulate 
 * 2021-06-08 [0f7a5e5](http://github.com/hugsy/gef/commit/0f7a5e5ce2f7724eb27a0f987d3670d9fb672ffd) &bull; *Michael Hoefler* &bull; fix alias tests 
 * 2021-06-09 [bd8e338](http://github.com/hugsy/gef/commit/bd8e3387c61675b7b2a1e9c961e50fb8c57593a0) &bull; *hugsy* &bull; * improved command argparsing * ported a few functions to use it * updated unit tests to reflect the syntax change 
 * 2021-06-09 [08946bd](http://github.com/hugsy/gef/commit/08946bda086e3edac047e97daf48b5488543bae3) &bull; *hugsy* &bull; - switched `emulate` to use the new argparser 
 * 2021-06-09 [34b15d6](http://github.com/hugsy/gef/commit/34b15d63face971c0fb0bae78f545846a3c1327c) &bull; *hugsy* &bull; implemented argparsing using argparse, applied on unicorn-emulate 
 * 2021-06-09 [6482690](http://github.com/hugsy/gef/commit/64826905f8cbcc67cc7cca4be42347900f4fc7c4) &bull; *hugsy* &bull; * improved command argparsing * ported a few functions to use it * updated unit tests to reflect the syntax change 
 * 2021-06-09 [2892d15](http://github.com/hugsy/gef/commit/2892d15382168927d0f480753036ba8358a9ef47) &bull; *hugsy* &bull; - switched `emulate` to use the new argparser 
 * 2021-06-09 [0b503b5](http://github.com/hugsy/gef/commit/0b503b55e9c187475f4acbdf42e63f1755d38ea3) &bull; *hugsy* &bull; merge fix 
 * 2021-06-10 [bd65e81](http://github.com/hugsy/gef/commit/bd65e811f0baeb90197f0a276b4f067b5b1cd19e) &bull; *hugsy* &bull; [argparse] allowing positional arguments to pass default values as list 
 * 2021-06-11 [d6fd038](http://github.com/hugsy/gef/commit/d6fd038a4e768027d7cbd33ac38279f40a742982) &bull; *hugsy* &bull; New addition for #659 : - added `pattern` command - changed the default period to 4 for better compat with `pwntools` (fixed #658) - updated unit tests 
 * 2021-06-12 [bbe84db](http://github.com/hugsy/gef/commit/bbe84db62dff7cbce96d0ef7a17c66b258f6aef5) &bull; *hugsy* &bull; Added argparsing support and unit test  for processlisting  (#659) 
 * 2021-06-12 [782dd88](http://github.com/hugsy/gef/commit/782dd88bb6b285d107fae9368ee2f06b5aacddb1) &bull; *hugsy* &bull; - added RemoteCommand for argparsing (#659) - fixed ArgumentParser prog name (#659) 
 * 2021-06-14 [2ed2373](http://github.com/hugsy/gef/commit/2ed237378b77f2b3483e0e9be00f7310d30ce501) &bull; *Michael Hoefler* &bull; gef command documentation 
 * 2021-06-14 [9e27f9e](http://github.com/hugsy/gef/commit/9e27f9ead0fa5afee4026af9751631bef9b6084c) &bull; *Michael Hoefler* &bull; fix period 
 * 2021-06-15 [d0a01da](http://github.com/hugsy/gef/commit/d0a01dafe3b730bf4947cdc3ae8142a474b59abb) &bull; *Michael Hoefler* &bull; gef subcommand documentation 
 * 2021-06-15 [50ea000](http://github.com/hugsy/gef/commit/50ea000a65a7c4769342e27af2a442f05003585a) &bull; *Michael Hoefler* &bull; fixes + line wrapping 
 * 2021-06-20 [5d167f8](http://github.com/hugsy/gef/commit/5d167f8674fad9782f570b1eabfeecd31a113d90) &bull; *hugsy* &bull; * Addresses most of the concerns raised in the review of PR #661 * Added the possibility to use short flags * Updated the documentation to provide relevant examples 
 * 2021-06-20 [7e47e8f](http://github.com/hugsy/gef/commit/7e47e8f0d19a63d627ccc5ecab60aa75a5d434aa) &bull; *hugsy* &bull; Updated example of `print-format` to reflect short flag 
 * 2021-06-20 [f52183f](http://github.com/hugsy/gef/commit/f52183ffe7e20c228c10f43f0d95a86caa3dbba2) &bull; *hugsy* &bull; [printf-format] fixed incorrect syntax 
 * 2021-06-20 [eda454b](http://github.com/hugsy/gef/commit/eda454b6c55ba166f6cc01f9c3dc4f387c24943a) &bull; *hugsy* &bull; Doc: added an example for `parse_arguments` to illustrate how to use positional arguments with a list of option (as suggested in #661) 
 * 2021-06-21 [abe837e](http://github.com/hugsy/gef/commit/abe837e2a88d4e5f9c988b1d46fd089d1d014752) &bull; *hugsy* &bull; Making @Grazfather happy again 😀(issue #661) 
 * 2021-06-22 [9f9b9b9](http://github.com/hugsy/gef/commit/9f9b9b9d977421d2e41e0c291e73600ce87919ae) &bull; *hugsy* &bull; [copy_to_clipboard] using context (see #661) 
 * 2021-06-22 [202a185](http://github.com/hugsy/gef/commit/202a185e50c1829cdd522c047793504b5e236131) &bull; *hugsy* &bull; Fixed issue #664 : adding an acknowledgement section to project sponsors 
 * 2021-06-22 [7e62dda](http://github.com/hugsy/gef/commit/7e62dda591770c9f61977aae4686abefee84450c) &bull; *hugsy* &bull; [gef-remote] fixed missing incorrect option name 
 * 2021-06-22 [a652786](http://github.com/hugsy/gef/commit/a652786b87d885e17b4101568178142ff2827b45) &bull; *hugsy* &bull; [capstone-disassemble] using new argparsing (see ##661) 
 * 2021-06-22 [5f5ba01](http://github.com/hugsy/gef/commit/5f5ba011640421afff3ebcc87bd1dee42275f362) &bull; *hugsy* &bull; [registers] using new argparsing (issue #661) 
 * 2021-06-22 [c5644b1](http://github.com/hugsy/gef/commit/c5644b196a3766c16c4bc84a87f2d0d72eddef70) &bull; *hugsy* &bull; [elf-info] using new argparsing (issue #661) 
 * 2021-06-22 [9ffe5cd](http://github.com/hugsy/gef/commit/9ffe5cd4e5241a4be51a93c501231b331111195a) &bull; *hugsy* &bull; [name-break] using new argparsing (issue #661) 
 * 2021-06-22 [e38f7dd](http://github.com/hugsy/gef/commit/e38f7dde4a390d217f82032bc49898a6e313a6f4) &bull; *hugsy* &bull; [hexdump] using new argparsing (issue #661) 
 * 2021-06-22 [556aa92](http://github.com/hugsy/gef/commit/556aa92ffbd080d1dff6dacffd79f9066fa164c5) &bull; *hugsy* &bull; [patch] using new argparsing (issue #661) 
 * 2021-06-23 [4661f81](http://github.com/hugsy/gef/commit/4661f819e6f59f0e0a789cedff9afc2179350397) &bull; *hugsy* &bull; [discord-notify] using embeds to make things more explicit, bumped to 0.3.2 
 * 2021-06-24 [a6d2baf](http://github.com/hugsy/gef/commit/a6d2baf899cbb065976b23feb144bb175c667c02) &bull; *hugsy* &bull; [discord-notify] fixed json typo + removed unsupported hyperlink 
 * 2021-06-24 [b335d3d](http://github.com/hugsy/gef/commit/b335d3dbb7eca71316565bca9740e74f0ea455ac) &bull; *hugsy* &bull; [discord-notify] better formatting using embeds 
 * 2021-06-24 [de053b8](http://github.com/hugsy/gef/commit/de053b88dc25b955b16efc670d8a15a8bd4e8728) &bull; *hugsy* &bull; [discord-notify] fixed typo 
 * 2021-06-24 [6acab61](http://github.com/hugsy/gef/commit/6acab6132489b095c7675fb62b89add5c5bd5a43) &bull; *hugsy* &bull; [discord-notify] fixed syntax error 
 * 2021-06-24 [417f7b0](http://github.com/hugsy/gef/commit/417f7b0d2ac772e36ca2a9a7ef6fc0a9c2b10318) &bull; *hugsy* &bull; [discord-notify] using the proper syntax for `Ilshidur/action-discord` 
 * 2021-06-24 [71dbadd](http://github.com/hugsy/gef/commit/71dbaddac3ae2bf4f46a968c60d9aa8ea43b422c) &bull; *hugsy* &bull; [discord-notify] typo 
 * 2021-06-25 [762e73d](http://github.com/hugsy/gef/commit/762e73df4ac1ef016f7e9335c8ba6a3eaaaaf93a) &bull; *Boris Molodenkov* &bull; Restore selected frame after context_threads 
 * 2021-06-24 [7b2a0a2](http://github.com/hugsy/gef/commit/7b2a0a265937ef738669a22ccf6a082cbc8dff7a) &bull; *hugsy* &bull; [discord-notify] bye bye trash `Ilshidur/action-discord` , hello `sarisia/actions-status-discord` 
 * 2021-06-24 [8b04ca3](http://github.com/hugsy/gef/commit/8b04ca382a55d2cced86446de97302d69d35de5f) &bull; *hugsy* &bull; [discord-notify] syntax 
 * 2021-06-24 [847949b](http://github.com/hugsy/gef/commit/847949b7547fac856b008a30ad493abc2ef49291) &bull; *hugsy* &bull; [discord-notify] last touch 
 * 2021-06-24 [ba88941](http://github.com/hugsy/gef/commit/ba889410af0dba8994d115ba0bdb3d0404beec8d) &bull; *hugsy* &bull; [discord-notify] last touch 
 * 2021-06-25 [ee3564a](http://github.com/hugsy/gef/commit/ee3564aaf3c540e9ea84de8473f8c16a4953c94d) &bull; *hugsy* &bull; [discord-notify] fixed syntax error in issue notification 
 * 2021-06-26 [0749012](http://github.com/hugsy/gef/commit/0749012893b26a0eb9a02fa1bbd30e646f753b6c) &bull; *Boris Molodenkov* &bull; Fix incorrect lru_cache usage 
 * 2021-06-26 [ac545fe](http://github.com/hugsy/gef/commit/ac545fe3bcfd62e1cc1531bfcc8f70700aa6ce03) &bull; *hugsy* &bull; [ci] caching libs for test runs - #660 
 * 2021-06-26 [527b840](http://github.com/hugsy/gef/commit/527b840ee6a52bfc1008941ef6c6b8cf60ca5bbd) &bull; *hugsy* &bull; [ci] better jobs splitting - #660 
 * 2021-06-26 [e83dafc](http://github.com/hugsy/gef/commit/e83dafc25c1abda8fa4d18425eba67aa90bb0156) &bull; *hugsy* &bull; [ci] fixed yaml syntax error - #660 
 * 2021-06-27 [75223e5](http://github.com/hugsy/gef/commit/75223e50662eb5b778573020fad69748b2bca914) &bull; *Boris Molodenkov* &bull; Fix frame level in context_trace and highlight selected frame 
 * 2021-06-28 [cdb4896](http://github.com/hugsy/gef/commit/cdb48960ec04312a38d77309492a5455abbff9d6) &bull; *Boris Molodenkov* &bull; Fix context regs order (#670) 
 * 2021-06-30 [15ffa33](http://github.com/hugsy/gef/commit/15ffa335a42c0cd6de16604d56bc38e6f6564a8e) &bull; *hugsy* &bull; Improvement of `get_register` after #669 to restore some decent level of caching 
 * 2021-06-30 [cbaf9d7](http://github.com/hugsy/gef/commit/cbaf9d728d18daec3a1c5816ef2bb22c2de3a2dd) &bull; *hugsy* &bull; Merge branch 'dev' of github.com:hugsy/gef into dev 
 * 2021-06-30 [42c7e51](http://github.com/hugsy/gef/commit/42c7e51734bc57f58f56b327b7dd54722ae7f3d4) &bull; *hugsy* &bull; added regression tests for the PRs #670 and #668 
 * 2021-06-30 [62b7a04](http://github.com/hugsy/gef/commit/62b7a04bd9ba4851664be38c19307148504d9194) &bull; *hugsy* &bull; added regression tests for the PRs #670 and #668 
 * 2021-06-30 [1e9b7de](http://github.com/hugsy/gef/commit/1e9b7deb9164576098104442b68a2d323e775ede) &bull; *hugsy* &bull; removed useless todo 
 * 2021-06-30 [686ff82](http://github.com/hugsy/gef/commit/686ff82f0e8d836e1397ca5435d40169f5b33536) &bull; *hugsy* &bull; fixed pylint 
 * 2021-07-01 [39503df](http://github.com/hugsy/gef/commit/39503dfdb48d248a8f04deda094fa9989fc3759e) &bull; *hugsy* &bull; turn pylint into something actionable 
 * 2021-07-01 [4d4a813](http://github.com/hugsy/gef/commit/4d4a813cccdba56d96a7650172918fc7c6124f5d) &bull; *hugsy* &bull; minor pylint suggested improvements 
 * 2021-07-01 [ee81a62](http://github.com/hugsy/gef/commit/ee81a62026e74315e6ad4205688dbf916df18e25) &bull; *hugsy* &bull; minor pylint fixes (`no-space-after-comma`) 
 * 2021-07-01 [42d524c](http://github.com/hugsy/gef/commit/42d524ca059ff3b0138e6008d1aff4666985f3e0) &bull; *hugsy* &bull; [pylint]  add missing-import exceptions 
 * 2021-07-01 [05d244f](http://github.com/hugsy/gef/commit/05d244ff8f797ad68cbd2b2580087ad944683f4e) &bull; *hugsy* &bull; [actions] fixing gh action cache 
 * 2021-07-01 [6ba3867](http://github.com/hugsy/gef/commit/6ba38675b037a0421bf85c40678ecac0d2ebb23d) &bull; *hugsy* &bull; [actions] fixed missing path for gh action cache 
 * 2021-07-01 [185fa0c](http://github.com/hugsy/gef/commit/185fa0c64b9e02a3508b6212e21892badd55ceb0) &bull; *hugsy* &bull; [actions] fixed bad archive path 
 * 2021-07-01 [2b5bb29](http://github.com/hugsy/gef/commit/2b5bb296796e17f234871c27b4b4d38d2f605f9b) &bull; *hugsy* &bull; [actions] caching is hard 
 * 2021-07-01 [5fb2e90](http://github.com/hugsy/gef/commit/5fb2e90b723d2dc2679b641fe3625ac01f36f1eb) &bull; *hugsy* &bull; adding `parse_arguments` for `pie *` commands (#659) 
 * 2021-07-03 [c9b0bda](http://github.com/hugsy/gef/commit/c9b0bda450864efe388d10021554ec9bde6ccc11) &bull; *hugsy* &bull; Merge branch 'dev' of github.com:hugsy/gef into dev 
 * 2021-07-03 [ab0edc5](http://github.com/hugsy/gef/commit/ab0edc52b5d859b34ffc431459a5604bd45b75e7) &bull; *hugsy* &bull; [ci] Fixed a bug in`run-test.yml` that prevented cache to be correctly collected 
 * 2021-07-05 [7d1ab00](http://github.com/hugsy/gef/commit/7d1ab00d539b22070fa2c49dbbcf19bde0727b98) &bull; *Boris Molodenkov* &bull; Add missing aarch64 condition codes (#671) 
 * 2021-07-05 [80c768a](http://github.com/hugsy/gef/commit/80c768aaab6ed864c55940741343683ac73e6462) &bull; *Grazfather* &bull; Remove more python2-specific stuff 
 * 2021-07-05 [e95b072](http://github.com/hugsy/gef/commit/e95b072f5684857ffd3d3c95c3369f83cdba0dee) &bull; *Grazfather* &bull; Fix caching in CI tests 
</details>

<hr /><em>This discussion was created from the release <a href='https://github.com/hugsy/gef/releases/tag/2021.07'>2021.07 - Impractical Information</a>.</em>

**Comment by ammarfaizi2** [comment]
Cc: @hugsy
I believe something went wrong with the commit details formatting. Screenshot attached.
![Screenshot_2021-07-06_17-07-54](https://user-images.githubusercontent.com/26004054/124582953-bb9ef580-de7c-11eb-832c-cbe050d8e0ea.png)

**Reply by hugsy** [reply]
Fixed on the release (and in the script!), thanks!
